File: Atracciones.Frontend/atraxia/src/app/services/atracciones.service.ts
import { Injectable } from '@angular/core';
import api from './api.service';
import axios from 'axios';
import { environment } from '../../environments/environment';

export interface ResenaItem {
  clienteId: number;
  atraccionId: number;
  calificacion: number;
  comentario: string;
  fecha: string;
}

export interface AtraccionItem {
  id: string; nombre: string; ciudad: string; pais: string;
  tipo_nombre: string; descripcion_corta: string; precio_desde: number;
  moneda: string; calificacion: number; total_resenias: number;
  duracion_minutos: number; imagen_principal: string; etiquetas: string[];
  disponible: boolean; disponible_hoy: boolean;
  proxima_fecha_disponible: string; cupos_disponibles: number;
  idiomas_disponibles?: string[];
}

export interface AtraccionDetalle {
  id: string; nombre: string; descripcion: string; imagenes: string[];
  incluye: string[]; no_incluye: string[]; punto_encuentro: string | null;
  incluye_transporte: boolean; incluye_acompaniante: boolean;
  tickets: { horId: number; tck_guid: string; tipo: string; precio: number; moneda: string }[];
  horarios_proximos: { horarioId: number; hor_guid: string | null; atraccionId: number; fecha: string; hora_inicio: string; hora_fin: string; cupos: number }[];
}

export interface AtraccionPayload {
  destinoId: number;
  nombre: string;
  descripcion: string;
  direccion: string;
  duracionMinutos: number;
  puntoEncuentro: string;
  moneda: string;
  precioReferencia: number;
  incluyeTransporte: boolean;
  incluyeAcompaniante: boolean;
  categoriaIds: number[];
  idiomaIds: number[];
  incluyeIds: number[];
  noIncluyeIds: number[];
  tagIds: number[];
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  status: number;
  data: T;
}

function adminApi() {
  const token = localStorage.getItem('atraxia_admin_token');
  return axios.create({
    baseURL: environment.apiUrl,
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {})
    }
  });
}

@Injectable({ providedIn: 'root' })
export class AtraccionesService {
  async getAtracciones(filtros: Record<string, any> = {}) {
    const params: any = {};
    Object.entries(filtros).forEach(([k, v]) => { if (v !== '' && v !== undefined) params[k] = v; });
    const res = await api.get('/atracciones', { params });
    return res.data;
  }

  async getAtraccion(id: string): Promise<AtraccionDetalle> {
    const res = await api.get(`/atracciones/${id}`);
    return res.data.data;
  }

  async reservar(payload: any) {
    const res = await api.post('/reservas', payload);
    return res.data;
  }

  async confirmarPago(guid: string, payload: any = {}) {
    const res = await api.post(`/reservas/${guid}/pagos/confirmacion`, payload);
    return res.data;
  }

  async getResenas(id: string): Promise<ResenaItem[]> {
    const res = await api.get(`/atracciones/${id}/resenias`);
    return res.data.data;
  }

  async getHorariosByAtraccion(guid: string): Promise<any[]> {
    const res = await api.get(`/atracciones/${guid}/horarios`);
    return res.data.data;
  }

  async getTicketsPorHorario(guid: string, horarioGuid: string): Promise<any[]> {
    const res = await api.get(`/atracciones/${guid}/horarios/${horarioGuid}/tickets`);
    return res.data.data.items;
  }

  // Admin Methods
  async getInternalById(id: string): Promise<any> {
    const res = await adminApi().get<ApiResponse<any>>(`/atracciones/internal/${id}`);
    return res.data.data;
  }

  async create(payload: AtraccionPayload): Promise<ApiResponse<any>> {
    const res = await adminApi().post<ApiResponse<any>>('/atracciones', payload);
    return res.data;
  }

  async update(id: string, payload: AtraccionPayload): Promise<ApiResponse<any>> {
    const res = await adminApi().put<ApiResponse<any>>(`/atracciones/${id}`, payload);
    return res.data;
  }

  async delete(id: string): Promise<ApiResponse<any>> {
    const res = await adminApi().delete<ApiResponse<any>>(`/atracciones/${id}`);
    return res.data;
  }
}

File: Atracciones.Frontend/atraxia_mobile/lib/services/atracciones_service.dart
import 'dart:convert';
import '../model/atraccion.dart';
import 'api_service.dart';

class AtraccionesService {
  final ApiService _api = ApiService();

  Future<Map<String, dynamic>> getAtracciones({Map<String, dynamic>? filtros}) async {
    final params = <String, dynamic>{};
    filtros?.forEach((k, v) {
      if (v != null && v != '') {
        params[k] = v;
      }
    });
    if (params.containsKey('page')) {
      params['PageNumber'] = params['page'];
      params.remove('page');
    }
    if (params.containsKey('Disponible')) {
      params['Disponible'] = params['Disponible'] == 'true' || params['Disponible'] == true;
    }

    final res = await _api.get('/atracciones', queryParams: params);
    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body);
    }
    throw Exception('Error al cargar atracciones');
  }

  Future<AtraccionDetalle> getAtraccion(String id) async {
    final res = await _api.get('/atracciones/$id');
    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = jsonDecode(res.body);
      return AtraccionDetalle.fromJson(body['data']);
    }
    throw Exception('Error al cargar la atracción');
  }

  Future<Map<String, dynamic>> reservar(Map<String, dynamic> payload) async {
    final res = await _api.post('/reservas', payload);
    return jsonDecode(res.body);
  }

  Future<Map<String, dynamic>> confirmarPago(String guid, Map<String, dynamic> payload) async {
    final res = await _api.post('/reservas/$guid/pagos/confirmacion', payload);
    return jsonDecode(res.body);
  }

  Future<List<dynamic>> getResenas(String id) async {
    final res = await _api.get('/atracciones/$id/resenias');
    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = jsonDecode(res.body);
      return body['data'] ?? [];
    }
    return [];
  }

  Future<List<HorarioProximo>> getHorariosByAtraccion(String guid) async {
    final res = await _api.get('/atracciones/$guid/horarios');
    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = jsonDecode(res.body);
      final list = body['data'] as List?;
      return list?.map((e) => HorarioProximo.fromJson(e)).toList() ?? [];
    }
    return [];
  }

  Future<List<TicketInfo>> getTicketsPorHorario(String guid, String horarioGuid) async {
    final res = await _api.get('/atracciones/$guid/horarios/$horarioGuid/tickets');
    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = jsonDecode(res.body);
      final list = body['data']?['items'] as List?;
      return list?.map((e) => TicketInfo.fromJson(e)).toList() ?? [];
    }
    return [];
  }

  Future<AtraccionDetalle> getAtraccionCompleta(String guid) async {
    const String query = '''
      query GetAtraccionCompleta(\$guid: String!) {
        atraccion(guid: \$guid) {
          id
          nombre
          descripcion
          imagenes
        }
        horarios(guid: \$guid) {
          horGuid
          fecha
          horaInicio
          horaFin
          cupos
        }
      }
    ''';

    final res = await _api.queryGraphQL('/graphql/atraccion', query, variables: {'guid': guid});
    if (res.hasException) throw Exception(res.exception.toString());
    
    final data = res.data;
    if (data == null) throw Exception('No data');

    final atr = AtraccionDetalle.fromJson(data['atraccion']);
    final hList = data['horarios'] as List?;
    atr.horariosProximos = hList?.map((e) => HorarioProximo.fromJson(e)).toList() ?? [];
    return atr;
  }

  Future<List<TicketInfo>> getTicketsGraphQL(String guid, String horarioId) async {
    const String query = '''
      query GetTickets(\$guid: String!, \$horarioId: String!) {
        tickets(guid: \$guid, horarioId: \$horarioId) {
          tckGuid
          tipo
          precio
          moneda
        }
      }
    ''';

    final res = await _api.queryGraphQL('/graphql/atraccion', query, variables: {
      'guid': guid,
      'horarioId': horarioId
    });
    if (res.hasException) throw Exception(res.exception.toString());
    
    final list = res.data?['tickets'] as List?;
    return list?.map((e) => TicketInfo.fromJson(e)).toList() ?? [];
  }
}
